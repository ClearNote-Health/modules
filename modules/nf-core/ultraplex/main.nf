process ULTRAPLEX {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/ultraplex:1.2.5--py38h4a8c8d9_0' :
        'biocontainers/ultraplex:1.2.5--py38h4a8c8d9_0' }"

    input:
    tuple val(meta), path(reads)
    path(barcode_file)
    val(adapter_seq)

    output:
    tuple val(meta), path("*_matched.fastq.gz")    , emit: fastq
    tuple val(meta), path("*_no_match_*.fastq.gz") , emit: no_match_fastq, optional: true
    path "*.log"                                   , emit: report
    path "versions.yml"                            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def VERSION = "1.2.5" // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
    def args    = task.ext.args ?: ''
    prefix      = task.ext.prefix ?: "${meta.id}"

    def adapter_seq_command = ''
    if(adapter_seq) {
        adapter_seq_command = "--adapter ${adapter_seq}"
    }

    read_list = reads.collect{it.toString()}
    if (read_list.size > 1){
        ultraplex_command = """ultraplex \\
        --inputfastq ${read_list[0]} \\
        --input_2 ${read_list[1]} \\
        --barcodes $barcode_file \\
        --threads $task.cpus \\
        --outputprefix ${prefix} \\
        $args \\
        $adapter_seq_command"""
    } else {
        ultraplex_command = """ultraplex \\
        --inputfastq ${read_list[0]} \\
        --barcodes $barcode_file \\
        --threads $task.cpus \\
        --outputprefix ${prefix} \\
        $args \\
        $adapter_seq_command"""
    }

    """
    ${ultraplex_command}

    ## rename the matched files to be able to emit only the matched files
    MATCHES=\$( find . -type f -name '*.fastq.gz' ! -name '*_no_match_*.fastq.gz' )
    for MATCH in \$MATCHES; do
        mv \$MATCH \${MATCH/.fastq.gz/_matched.fastq.gz}
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ultraplex: $VERSION
    END_VERSIONS
    """

    stub:
    def VERSION = "1.2.5" // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
    def args    = task.ext.args ?: ''
    prefix      = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ultraplex_${prefix}_Sample1_Fwd_matched.fastq.gz
    echo "" | gzip > ultraplex_${prefix}_Sample1_Rev_matched.fastq.gz
    echo "" | gzip > ultraplex_${prefix}_Sample2_Fwd_matched.fastq.gz
    echo "" | gzip > ultraplex_${prefix}_Sample2_Rev_matched.fastq.gz
    echo "" | gzip > ultraplex_${prefix}_no_match_Fwd.fastq.gz
    echo "" | gzip > ultraplex_${prefix}_no_match_Rev.fastq.gz

    touch ultraplex.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ultraplex: $VERSION
    END_VERSIONS
    """
}
