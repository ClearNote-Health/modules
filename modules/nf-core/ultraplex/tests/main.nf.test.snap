{
    "ultraplex - incorrect barcodes": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "ultraplex_demux_Sample2_Fwd_matched.fastq.gz:md5,4451dd878b7d12ff8ceb05fce2ebfa89",
                        "ultraplex_demux_Sample2_Rev_matched.fastq.gz:md5,f1c6ca8d7f8407b5bb0b6e8906fe41ad"
                    ]
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-05T06:08:12.106448378"
    },
    "ultraplex - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "ultraplex_demux_Sample1_matched.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "ultraplex_demux_Sample2_matched.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "ultraplex_demux_Sample1_no_match.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "ultraplex_demux_Sample2_no_match.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "2": [
                    "ultraplex.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "3": [
                    "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
                ],
                "fastq": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "ultraplex_demux_Sample1_matched.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "ultraplex_demux_Sample2_matched.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "no_match_fastq": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "ultraplex_demux_Sample1_no_match.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "ultraplex_demux_Sample2_no_match.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "report": [
                    "ultraplex.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "versions": [
                    "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-05T05:52:36.211334997"
    },
    "ultraplex - adaptor": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "ultraplex_demux_Sample1_matched.fastq.gz:md5,77870ddf54ed3e474d4bb63fea65fc08",
                        "ultraplex_demux_Sample2_matched.fastq.gz:md5,f3960ec2757672915538923ecb3539f8"
                    ]
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-05T05:54:27.606493007"
    },
    "ultraplex": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "ultraplex_demux_Sample1_matched.fastq.gz:md5,77870ddf54ed3e474d4bb63fea65fc08",
                        "ultraplex_demux_Sample2_matched.fastq.gz:md5,f3960ec2757672915538923ecb3539f8"
                    ]
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-05T05:54:11.263761896"
    },
    "ultraplex - pairedend": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    [
                        "ultraplex_demux_Sample1_Fwd_matched.fastq.gz:md5,6b516012c4b19bb0d2109881e3a36102",
                        "ultraplex_demux_Sample1_Rev_matched.fastq.gz:md5,79637dfeca38bc9314cc5bde37373b5f",
                        "ultraplex_demux_Sample2_Fwd_matched.fastq.gz:md5,1f17530c291f74a6b29c63ebaba15c96",
                        "ultraplex_demux_Sample2_Rev_matched.fastq.gz:md5,81ea922ee9842b8f6fe9654406b4f479"
                    ]
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,916995657ca9c5fd70171d69c3aba409"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-05T05:54:44.157822884"
    }
}